source.go

```diff
@@ -8,11 +8,147 @@ import (
 	"strings"
 
 	"github.com/moby/buildkit/client/llb"
+	"github.com/moby/buildkit/util/bklog"
 	"github.com/moby/buildkit/util/gitutil"
 	"github.com/pkg/errors"
 )
 
-type LLBGetter func(sOpts SourceOpts, opts ...llb.ConstraintsOpt) (llb.State, error)
+type FilterFunc = func(string, []string, []string, ...llb.ConstraintsOpt) llb.StateOption
+
+var errNoSourceVariant = fmt.Errorf("no source variant found")
+
+func GetSource(src Source, name string, filter FilterFunc, sOpt SourceOpts, opts ...llb.ConstraintsOpt) (llb.State, bool, error) {
+	// load the source
+	switch {
+	case src.HTTP != nil:
+		st, err := src.HTTP.AsState(name, opts...)
+		if err != nil {
+			return llb.Scratch(), false, err
+		}
+
+		return st.With(filter(src.Path, src.Includes, src.Excludes, opts...)), true, nil
+	case src.Git != nil:
+		st, err := src.Git.AsState(opts...)
+		if err != nil {
+			return llb.Scratch(), false, err
+		}
+		return st.With(filter(src.Path, src.Includes, src.Excludes)), true, nil
+	case src.Context != nil:
+		st, err := src.Context.AsState(&src, sOpt, opts...)
+		if err != nil {
+			return llb.Scratch(), false, nil
+		}
+		return st.With(filter(src.Path, []string{}, []string{})), true, nil
+
+	case src.DockerImage != nil:
+		st, err := src.DockerImage.AsState(name, &src, sOpt, opts...)
+		if err != nil {
+			return llb.Scratch(), false, err
+		}
+
+		var pathHandledByFilter bool
+		if src.DockerImage.Cmd != nil {
+			st = st.With(filter("/", src.Includes, src.Excludes))
+			pathHandledByFilter = false
+		} else {
+			pathHandledByFilter = true
+			st = st.With(filter(src.Path, src.Includes, src.Excludes))
+		}
+		return st, pathHandledByFilter, nil
+
+	case src.Build != nil:
+		st, err := src.Build.AsState(name, &src, filter, sOpt, opts...)
+		if err != nil {
+			return llb.Scratch(), false, err
+		}
+
+		return st.With(filter(src.Path, src.Includes, src.Excludes)), true, nil
+	}
+
+	return llb.Scratch(), false, errNoSourceVariant
+}
+
+func (src *SourceContext) AsState(parent *Source, sOpt SourceOpts, opts ...llb.ConstraintsOpt) (llb.State, error) {
```

Comment:
Can we pass the includes/excludes directly ins5tead of the full parent source?

Reply:
sure. we will also need to pass path in as well